### DISCUSSION TITLE
Development | Release v0.21 - 05/06/21

### DISCUSSION THREAD

**Discussion by johanjanssens** [joomlatools/joomlatools-pages#738] [Announcements]
Development of Pages is steaming ahead with the release of Pages 0.21.  Lots of changes in this release I'm very happy and excited about. 

- [Overview](#Overview)
- [Releases](#Releases)
- [Changelog](#Changelog)
- [Breaking](#breaking)

<a name="Overview"></a>
# Overview

This release concludes all the template engine work, there are a number of [breaking changes](#Breaking) that I all included in this release. The changes are all simple find/replace so nothing too hard. The template and frontmatter syntax is now considered stable.

## Extension API complete

A lot of work has been done to make extensions [first-class citizens](https://en.wikipedia.org/wiki/First-class_citizen) and make them very easy to create, update and distribute. With this release extensions can now be distributed as phar archives, and can be installed and updated from a URL. 

### Phar support 

[Phar](https://www.php.net/manual/en/book.phar.php) support has been on my wishlist for quite some time. You can bundle an extension as a single phar zip file, and just upload the file to the `/extensions` folder, no need to unzip it, it will just work!

![116628810-37963200-a950-11eb-8d50-c00fde863e7a](https://user-images.githubusercontent.com/266640/120901056-c590c700-c638-11eb-8089-24c062165856.png)

A phar is a zip so you can un-package and modify an extension easily if needed. Pages will also (re)-package them automatically for you so you do not need any additional build tools to go from code to archive.

### Historical context

Long ago during the development of Mambo 4.6 (which we never released) we worked on the concept of a unified installer. The idea was that everything would be a component, and could be installed as a single package. We implemented the unified installer in Joomla 1.5 but we never managed to make an extension into a single package.

Making this happen is not an easy feat. You need an architecture that can unify all extension types into a single type, which Joomla cannot do,  this is where Joomlatools Framework comes in. Combine this with the Phar support in PHP, add a specialised auto-loader and single file extension bundles are a thing.

## Joomla extension released

The Joomla extension offers a complete integration with Joomla, and is the first official extension for Pages that is being released. 

Extensions that are part of the Pages repo, are packaged when a a release is build and available through our our own Joomlatools  CDN powered by CloudFlare. Example: https://files.joomlatools.xyz/pages@0.21.1/extension-joomla.zip Installing an extension is easy  just specify the url in the [`/extensions/package.php`](https://github.com/joomlatools/joomlatools-pages/blob/master/contrib/sites/blog/extensions/package.php) file and Pages will install it for you on the fly. 

The Joomla extension offers support for content, modules and menus, with very powerful custom fields filtering allowing you to filter your Joomla articles with a simple syntax without needing to write any database queries or PHP code. 

### Examples

```php
<?php
$articles = collection('ext:joomla.model.articles', [
    'tags'  => ['foo', 'bar,quuz'],
    'field' => ['select' => ['A', 'B,C']],
]);
```

Get all articles which have `tags foo  AND bar OR quuz AND have a field select with value A AND B OR C` 

```php
<?php
$articles = collection('ext:joomla.model.articles', [
    'tags'  => ['foo', 'bar'],
    'field' => ['baz' => 'quux', 'select' => 'A,B'],
    'published' => 1,
    'sort' => 'created',
    'order' => 'desc',
]);
```

Get all articles that have `tags foo AND bar AND have a field baz with value quux AND a field select that has value A OR B  AND that are published, sort by created date and order descending`.

## Using Pages standalone

Pages is no longer a Joomla extension and thus Joomla derivative work. Pages it is now fully running standalone, the same codebase can both be installed and run in Joomla as be run standalone.

A [bootstrapper script](https://github.com/joomlatools/joomlatools-pages/blob/master/resources/pages/bootstrapper.php) is provided for, together with [support for the PHP build in server](https://github.com/joomlatools/joomlatools-pages/blob/master/resources/pages/server.php), making it super easy to develop with Pages, all you need is PHP 7.4 (no more Apache or MySQl) and you are good to go. 

### Historical context
 
The idea goes back to the GPL debate of 2007. _What if you could create your own component framework that doesn't derive from Joomla?_ It's one of things we got threatened with when we made it clear that all Joomla extensions are derivative works and need to be licensed GPL. People said they would do their own frameworks  ....

This question has always fascinated me, _is this actually possible_? Creating a component framework that is not a derivative work? The best way to proof this is to build something that can work both when installed in Joomla and can be run without any Joomla dependencies. If it can run in both contexts it's not derivative of one context and is considered a complete separate work. More then a decade later I'm happy to say YES! this is possible, but far from easy! 

This also means that **Pages is no longer GPL because it needs to be, it's GPL because I choose it to be**. 

## Superb Performance

Pages is fast, very fast, with caching enabled, even without using static caching Pages can handle requests at **+/- 40-50msec** when running standalone. Even when installed in Joomla you can still run Pages standalone on demand. You don't need to install a separate instance. Here is a little overview of possible setup configurations:

### 1.Joomla setup

Automatically runs inside the Joomla application context, connect to the Joomla database and session.  

> Can also be run standalone this way, by bootstrapping it! So you don't need two installs of Pages, you can install it in Joomla and bootstrap it to run it standalone.

### 2. Standalone setup

Runs in it's own application context, doesn't run a database connection or a php session unless one is setup.

> Can also be connected to Joomla this way, by setting up a database connection for the specific Joomla database.

The benefit of running Pages standalone is speed, even without caching Pages becomes 3x faster, pages requests go down to 150msec for generating a page and 50msec for revalidation a generated page from cache. Thats incredibly fast.

You get full flexibility, you can for example develop inside a Joomla context, and then run Pages standalone for your live site... etc.

##  Blog sample site

On popular request released a simple blog sample site that works with the Joomla blog sample data. Just unzip the package and you are ready to explore and play with Pages.

![Screenshot_2021-02-22 Pages Joomla Content](https://user-images.githubusercontent.com/266640/120902641-c7ab5380-c641-11eb-88c7-7e25ebd94e60.png)

Tried to make it as easy as possible to get started. Once unzipped Pages will automatically install the Joomla extension for you, no manual extra steps are required, and you are good to go. 

<a name="Releases"></a>
# Releases

Released v0.21.0, v0.21.1 More info: 

- v0.21.0: https://github.com/joomlatools/joomlatools-pages/releases/tag/v0.21.0
- v0.21.1: https://github.com/joomlatools/joomlatools-pages/releases/tag/v0.21.1

<a name="Changelog"></a>
# Changelog 

### 1. Template blocks #704

Implemented support for template blocks. Blocks are used for template inheritance and act as placeholders and replacements at the same time. All the block tag does is to tell the template engine that a child template may override those portions of the template.

```html
<ktml:block name="header">
      <header>
            <ktml:block>
     </header>
</ktml:block>
````

and to inject content into this block:

```html
<ktml:block extend="header">
      <?= partial('header) ?>
</ktml:block>
```

### 2. Frontmatter in partials #703

Added support for frontmatter in partials. This allows to use frontmatter in a partial to define default values for variables used in the partial, at the same time the frontmatter makes it easier to see what variables a partial accepts.

```php
---
foo: bar
joe: doe
---

<?= $foo >,<?= $joe >
```

### 3. Page attributes #700

Added support for page attributes. Attributes are used for page internal variables like `@route`, `@layout`, `@collection`, `@form`, ...

Page attributes are implicit, meaning that you do not need to add the `@`. This makes this change fully backwards compatible.  It also means that `@layout`and `layout`are the same variable. They are handled different internally but referenced in the same way as `page()->layout`

```yaml
---
@layout: /articles
@route: 'articles/[\d+:page]?'
@collection:
    type: ext:joomla.model.articles
    state:
        published: 1
         category: [foo, bar]
        sort: title

title: Articles
summary: My articles page
---
```

### 4. Improve template functions #356

The core template functions have been moved into separate files and new `snippet()`, `user()`, `element()` functions have been added. 

### 2. Add template_debug config option #678

Added a new `template_debug` config option to make it possible to enable/disable the template debug mode separately. By default Pages will put itself in debug model if you turn Joomla debug on, you can turn debug on your self using the global debug config option or you can enable/disable debug for template, and http client separately.

### 10. Simplify form honeypot #433

Improved the honeypot As part of this change the honeypot field no longer needs to be passed to the helper.

````php
<?= helper('form.honeypot'); ?>
````

### 1. Extension Management #689

Implemented support for installing, updating and archiving extensions. Installing of extensions is supported for valid phar zip archives, from path or url. 

```php
<?php
return [
    'http://example.com/foo.zip',
     '/path/to/bar.zip',
];
```

Pages can now also automatically update and archive installed extensions. Offering complete and easy to use extension management. 

```
21:05:09 08:57:02 - Install Success: http://joomla.test/joomla.zip?newest, version 1.0.0
21:05:09 08:57:13 - Update Success: http://joomla.test/joomla.zip?new, version 1.0.0
21:05:09 08:57:43 - Archive Updated: joomla.zip, version 1.0.0, build 1
```

### 2. Configure extension directory #725

Made it possible to configure the extension directory from which Pages loads extensions. You can now specify multiple paths, Pages will resolve them in FIFO order.

```php
'extension_path' =>
[
        PAGES_SITE_ROOT . '/extensions',
        KOOWA_ROOT.'/extensions',
],
```

In a multi-site setup this makes it possible to setup a directory from where Pages should load extensions that are used by all sites, while still allowing to have specific extensions per site.

### 3. Extension template locator #369

Implemented a template locator that can load both default templates and extension templates. This allows extension to provide re-usable templates, partials and layouts for use in Pages.  

```
---
layout: template://pages/document.html
---

<body>

<h1>Hello World</h1>

</body>
```

### 4. Move events subscribers location #687

The extensions event subscribers location to /extensions[name]/event/subscriber. Event subscribers are automatically included from this location. This PR extends upon the work done in: #144

### 5. Extension Router API #680

Allow extensions to provide their own custom router. This makes it possible for extensions to integrate with other routers or provide a complete custom routing.

```php
class ExtFooDispatcherRouter extends ComPagesDispatcherRouterAbstract
{
    public function getRoute($route, array $parameters = array())
    {
         $route = parent::getRoute($route, $parameters);
         //custom code here

        return $route;
    }
    
    public function generate($route = null, array $parameters = array())
    {
         //custom code here
        return parent::generate($route, $parameters);
    }
    
    public function resolve($route = null, array $parameters = array())
    {
        //custom code here
        return parent::resolve($route, $parameters);
    }
}
```

### 6. Automate release builds for extensions and sites #681

Joomlatools Pages release builds now automatically created packages for any `contrib/extensions` and `contrib/sites`packages. Extensions are build as phar archives. 

Packages can be downloaded from

- Github release assets, see https://github.com/joomlatools/joomlatools-pages/releases/tag/v0.21.1  
- Joomlatools CDN example, https://files.joomlatools.com/pages@0.21.1/extension-joomla.zip 

### 1. Joomla extension #363

Initial release of the Joomla extension which offers a tighter and specialised integration of Pages with Joomla Following features are available:

- collections based on com_content (articles, categories, fields and menus) with custom field support and tags and fields filtering
- simplified Joomla event handling (without needing plugins)
- the decorator is now part of this extension

````php
---
collection:
    model: ext:joomla.model.articles
    state:
        published: 1
         category: [foo, bar]
        sort: title
---

<ul>
<? foreach(collection() as $article): ?>
	<li><a href="<?= $article->getRoute() ?>"><?= $article->title; ?></a></li>
<? endforeach ?>
</ul>
````

### 2. Move module and plugins filter to Joomla extension #706

The plugins and module template filters have been moved into the Joomla extension to further decouple Pages from specific Joomla API's.

### 3. Improve collection hash calculation #719

Fixed an issue with the hash calculation for the joomla extension models. Also added a fallback from the menus model to be able to generate a hash for MySQL < 5.7.2 when Update_time is not available.

### 1. Allow Pages to run standalone #691

Refactored Pages to a Joomlatools Framework component instead of a Joomla component and makes it possible to run Pages standalone. 

Pages no longer requires Joomla to work. It can run both on it's own, without any Joomla dependencies, or can be installed in Joomla.

When installed in Joomla Pages is no longer installed in Joomla as a component. This means that Joomla no longer knows about the existence of Pages. Pages will not not show up in the list of installed extensions, nor in the menu item manager.

Pages is now bundled together with our Joomlatools Framework, when installed in Joomla uninstalling Joomlatools Framework will also remove the Pages component.

### 2. Add bootstrapper script #722

Added a bootstrapper script to make it very easy to bootstrap Pages when running standalone. 

```php
<?php
require dirname(__DIR__) .'/vendor/joomlatools/pages/resources/pages/bootstrapper.php';
```

### 3. Support for PHP built-in web server #723

Implements a server script that makes it possible to use the build in PHP server with Pages. The server script works in the same way as the rewrite rules for Apache. It offers support for `/theme` and `/images` folder aliasing, and handles rewrites for the Pages build in static cache.

```
> cd /web
> PHP_CLI_SERVER_WORKERS=10 PAGES_SITE=foo php -S localhost:8080 server.php 
```

### 1. Performance improvements #712

Implemented various performance improvements to increase page generation performanceInitial. Testing shows the following performance stats when running Pages standalone (measured on PHP 7.3, local setup)

- Generating a collection +/- **300-400msec** 
- Generating an item/page +/- **150-200msec**
- Revalidating the dynamic HTTP cache, with object cache validation: +/- **50msec**
- Revalidating the dynamic HTTP cache, without object cache validation: +/- **40msec** (*)
- Using the static HTTP cache: +/- **10msec** 

### 2. Improve page access handling #358

The page access logic has been consolidated in a new ComPagesRegistry::isPageAccessible() method. The method is called by the  dispatcher the pages model to check of a page is accessible.

### 3. Update php-markdown to v1.9 #717

The included php-markdown library has been updated to v1.9 to resolve compat issues with PHP8.

### 4. Improve exception handling #697

Made some smaller improvements to the exception handling. If debug is enabled Pages will generate a backtrace also for none html pages, eg, rss, xml, json.

### 5. Re-organise repo #676

Re-organised the repo, two main directories have been added:

- `contrib`: Contains all official contributions for Pages (extensions, sites, themes, partials, ....)
- `resources`: Contains all external resources, mason, prefetcher, proxy etc.

All code is now maintained and developed together to ensure everything is kept in sync with feature improvements and changes made in Pages.

### 6. Blog sample site #671

Released a simple example site and theme showcasing how to use the Joomla content extension for Pages.

<a name="breaking"></a>
### 1. Page route improvements #709

Improved the consistency of route path, custom routes now need to start with `/`. 

```yaml
---
@route: /blog/[:slug]
---
```

or 

```php
<?= route('/path/to/page') ?>
```

### 2. Collection naming conflicts #707

Improve the collection naming to prevent naming conflicts between custom collections and collections defined in pages. Before this change conflicts could happen between a custom collection for example pages and a page with the path /pages.

```yaml
@collection:
     extend: /path/to/page
```

or 

```php
<? collection('/path/to/page') ?>
```

 The pages collection folder state has been renamed to `path`. The path should now also be an absolute path, prefixed with `/`.
 
 ```yaml
@collection:
    state:
        path: /path/to/directory
```

### 3. Extension template locator #369

To improve consistency `layouts` and `partials`are now loaded from a root templates folder. Changing the default directory structure too:

````
/pages
  ├── about.html.php
  ├── products.html.php 
  ├── contact.html.php 
  └── index.html.php
/templates
   ├── /layouts
        ├── default.html.php
        └── products.html.php
   └── /partials
         ├── header.html.php
         ├── menu.html.php
         ├── product.html.php
         └── footer.html.php
/data
  ├── products.yaml
  └── footer.json
/theme
  ├── css
  ├── js
  ├── fonts
  └── images
/cache
config.php
````

As part of this change partial imports also need to be prefixed with '/'

```yaml
---
@layout: /path/to/layout
---
```

or 

```php
<?= import('/path/to/partial') ?>
```

#### `date()`

The `date()` function has been refactored and the arguments have been switched to follow it's [PHP counterpart](https://www.php.net/manual/en/function.date.php)

`date( string $format = '', [, mixed $date = 'now' ])`

### 5. Extension template locator #369

The templates provided by Pages have been moved: 

- `com:pages.collection.newsfeed.rss` => `template://pages/newsfeeds.rss`
- `com:pages.collection.sitemap.xml` => `template://pages/sitemap.xml`

**Comment by raramuridesign** [comment]
@johanjanssens Is the component recognised in Joomla Updater?

**Reply by johanjanssens** [reply]
If with component you mean Pages itself, then the answer is no. Pages is no longer a Joomla extension, its not 'installed' in joomla. I have auto-updates on the todo for a later release: https://github.com/joomlatools/joomlatools-pages/issues/211

Till then to update use following steps:

1. Uninstall Joomlatools Framework
2. Re-install Joomlatools Framework

Since all your data is in flat files, in the /joomlatools-pages folder uninstalling and re-installing has no side effects.

**Reply by raramuridesign** [reply]
@johanjanssens 
When I installed this on the project we added this by installing a component, then we added a Joomla extension.
So, when updating to new releases, how do we know which parts to update if new features come out?

**Reply by johanjanssens** [reply]
@raramuridesign Can you move this to a [new discussion](https://github.com/joomlatools/joomlatools-pages/discussions/categories/general) please and please elaborate a bit more, i'm not 100% sure what you mean. Happy to help you from there.

**Reply by raramuridesign** [reply]
https://github.com/joomlatools/joomlatools-pages/discussions/760
